src/analysis/processing/qgsalgorithmdownloadvectortiles.cpp

```diff
@@ -0,0 +1,225 @@
+  }
+
+  // count total number of tiles in the requested extent and zoom levels to see if it exceeds the tile limit
+  long long tileCount = 0;
+  QMap<int, QgsTileRange> tileRanges;
+  for ( int i = 0; i <= maxZoomLevel; i++ )
+  {
+    QgsTileMatrix tileMatrix = tileMatrixSet.tileMatrix( i );
+    QgsTileRange tileRange = tileMatrix.tileRangeFromExtent( extent );
+    tileRanges.insert( i, tileRange );
+    tileCount += ( tileRange.endColumn() - tileRange.startColumn() + 1 ) * ( tileRange.endRow() - tileRange.startRow() + 1 );
+    if ( tileCount > tileLimit )
+    {
+      throw QgsProcessingException( QObject::tr( "Requested number of tiles exceeds limit of %1 tiles. Please, select a smaller extent, reduce maximum zoom level or increase tile limit." ).arg( tileLimit ) );
+    }
+  }
+
+  std::unique_ptr<QgsMbTiles> writer = std::make_unique<QgsMbTiles>( outputFile );
+  if ( !writer->create() )
+  {
+    throw QgsProcessingException( QObject::tr( "Failed to create MBTiles file %1" ).arg( outputFile ) );
+  }
+  writer->setMetadataValue( "format", "pbf" );
+  writer->setMetadataValue( "name",  vtLayer->name() );
+  writer->setMetadataValue( "minzoom", QString::number( vtLayer->sourceMinZoom() ) );
+  writer->setMetadataValue( "maxzoom", QString::number( maxZoomLevel ) );
+  writer->setMetadataValue( "crs",  tileMatrixSet.rootMatrix().crs().authid() );
+  try
+  {
+    QgsCoordinateTransform ct( tileMatrixSet.rootMatrix().crs(), QgsCoordinateReferenceSystem( "EPSG:4326" ), context.transformContext() );
+    ct.setBallparkTransformsAreAppropriate( true );
+    QgsRectangle wgsExtent = ct.transform( extent );
+    QString boundsStr = QString( "%1,%2,%3,%4" )
+                        .arg( wgsExtent.xMinimum() ).arg( wgsExtent.yMinimum() )
+                        .arg( wgsExtent.xMaximum() ).arg( wgsExtent.yMaximum() );
+    writer->setMetadataValue( "bounds", boundsStr );
+  }
+  catch ( const QgsCsException & )
+  {
+    // bounds won't be written (not a problem - it is an optional value)
+  }
+
+  QgsProcessingMultiStepFeedback multiStepFeedback( maxZoomLevel + 1, feedback );
+
+  std::unique_ptr<QgsVectorTileLoader> loader;
+  QList<QgsVectorTileRawData> rawTiles;
+
+  QMap<int, QgsTileRange>::const_iterator it = tileRanges.constBegin();
+  while ( it != tileRanges.constEnd() )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    multiStepFeedback.setCurrentStep( it.key() );
+
+    QgsTileMatrix tileMatrix = tileMatrixSet.tileMatrix( it.key() );
+    tileCount = ( it.value().endColumn() - it.value().startColumn() + 1 ) * ( it.value().endRow() - it.value().startRow() + 1 );
+
+    const QPointF viewCenter = tileMatrix.mapToTileCoordinates( extent.center() );
+
+    long long tileNumber = 0;
+    if ( !provider->supportsAsync() )
```

Comment:
I don't think we need to worry about async handling here at all -- we could just always use the blocking method since we'll be running in a thread anyway.

Reply:
Should be fixed.